File: lesson-2_recursion-and-recurrence/the-longest-path/recursion/main.cpp
#include <iostream>
#include <vector>
#include <fmt/core.h>

int getTheLongest(std::vector<std::vector<int>> t, int n, int i, int j);

int main() {
    int n;
    std::cin >> n;

    auto t = std::vector<std::vector<int>>(n);
    for (int i = 0; i < n; i++) {
        t[i] = std::vector<int>(i + 1);
        for (int j = 0; j < i + 1; j++) {
            std::cin >> t[i][j];
        }
    }
    fmt::print("{}", getTheLongest(t, n, 0, 0));
    return 0;
}

int getTheLongest(std::vector<std::vector<int>> t, int n, int i, int j) {
    if (i == n - 1) {
        return t[i][j];
    }

    return std::max(getTheLongest(t, n, i + 1, j), getTheLongest(t, n, i + 1, j + 1)) + t[i][j];
}

File: lesson-2_recursion-and-recurrence/the-longest-path/recursion/main.go
package main

import (
    "fmt"
    "math"
)

func main() {
    var n int
    fmt.Scan(&n)

    t := make([][]int, n)
    for i := 0; i < n; i++ {
        t[i] = make([]int, i+1)
        for j := range t[i] {
            fmt.Scan(&t[i][j])
        }
    }
    fmt.Println(getTheLongest(t, n, 0, 0))
}

func getTheLongest(t [][]int, n, i, j int) int {
    if i == n-1 {
        return t[i][j]
    }
    return int(math.Max(float64(getTheLongest(t, n, i+1, j)), float64(getTheLongest(t, n, i+1, j+1)))) + t[i][j]
}
